apps/sms/test/marionette/mocks/mock_navigator_moz_mobile_message.js

```diff
@@ -27,215 +32,240 @@ Services.obs.addObserver(function(document) {
+    return getStorage().then(function(storage) {
+      var thread = getOrCreateThreadForRecipient(
+        storage,
         parameters.receiver || parameters.receivers[0]
-    );
-
-    var message = Cu.waiveXrays(Cu.cloneInto({
-      id: ++messageIdUniqueCounter,
-      iccId: null,
-      threadId: thread.id,
-      sender: null,
-      receivers: parameters.receivers,
-      receiver: parameters.receiver,
-      type: parameters.type,
-      delivery: 'sending',
-      deliveryInfo: [{
-        receiver: null,
-        deliveryStatus: 'not-applicable',
-        readStatus: 'not-applicable'
-      }],
-      subject: parameters.subject,
-      smil: parameters.smil,
-      body: parameters.body,
-      attachments: parameters.attachments,
-      timestamp: Date.now(),
-      sentTimestamp: Date.now(),
-      read: true
-    }, window));
-
-    thread.messages.push(message);
-
-    return message;
+      );
+
+      var message = {
+        id: ++storage.uniqueMessageIdCounter,
+        iccId: null,
+        threadId: thread.id,
+        sender: null,
+        receivers: parameters.receivers,
+        receiver: parameters.receiver,
+        type: parameters.type,
+        delivery: 'sending',
+        deliveryInfo: [{
+          receiver: null,
+          deliveryStatus: 'not-applicable',
+          readStatus: 'not-applicable'
+        }],
+        subject: parameters.subject,
+        smil: parameters.smil,
+        body: parameters.body,
+        attachments: parameters.attachments,
+        timestamp: Date.now(),
+        sentTimestamp: Date.now(),
+        read: true
+      };
+
+      thread.messages.push(message);
+
+      return message;
+    });
   }
 
   var MobileMessage = {
     getSegmentInfoForText: function(text) {
-      var request = Cu.waiveXrays(Cu.cloneInto({
-        onsuccess: null,
-        onerror: null
-      }, window));
-
-      window.setTimeout(function() {
-        var length = text.length;
-        var segmentLength = 160;
-        var charsUsedInLastSegment = (length % segmentLength);
-        var segments = Math.ceil(length / segmentLength);
-        if (typeof request.onsuccess === 'function') {
-          request.onsuccess.call(request, Cu.cloneInto({
-            target: {
-              result: {
-                segments: segments,
-                charsAvailableInLastSegment: charsUsedInLastSegment ?
-                  segmentLength - charsUsedInLastSegment : 0
-              }
-            }
-          }, window));
-        }
-      }, delayMs);
+      var request = Services.DOMRequest.createRequest(window);
+
+      var segmentLength = 160;
+      var charsUsedInLastSegment = (text.length % segmentLength);
+
+      // Very simple example of message segmentation logic, in reality it's more
+      // complex, e.g. see real "calculateLength" method here:
+      // https://android.googlesource.com/platform/frameworks/opt/telephony
+      // /+/android-5.0.1_r1/src/java/android/telephony/SmsMessage.java
```

Comment:
I was gonna say to point to B2G code instead (https://dxr.mozilla.org/mozilla-central/source/widget/android/AndroidBridge.cpp#1069) but it's really not readable...

Note that you can have the full URL on one line if there is no word before the URL, jshint should be happy with this.